frame/ibc/src/impls.rs

```diff
@@ -0,0 +1,752 @@
+
+			channel_key.extend_from_slice(channel_path.as_bytes());
+			inputs.push((channel_key.encode(), channel_end));
+			inputs.push((next_seq_ack_key.encode(), next_seq_ack.encode()));
+			inputs.push((next_seq_send_key.encode(), next_seq_send.encode()));
+			inputs.push((next_seq_recv_key.encode(), next_seq_recv.encode()));
+		}
+
+		// Insert packet commitments in trie
+		for ((port, channel, sequence), commitment) in PacketCommitment::<T>::iter() {
+			let channel_id =
+				channel_id_from_bytes(channel).map_err(|_| Error::<T>::DecodingError)?;
+			let port_id = port_id_from_bytes(port).map_err(|_| Error::<T>::DecodingError)?;
+			let sequence = ibc::core::ics04_channel::packet::Sequence::from(sequence);
+
+			let commitment_path = CommitmentsPath { port_id, channel_id, sequence };
+
+			let mut commitment_key = prefix.clone();
+			let commitment_path_str = format!("{}", commitment_path);
+			commitment_key.extend_from_slice(commitment_path_str.as_bytes());
+			inputs.push((commitment_key.encode(), commitment))
+		}
+
+		// Insert packet acknowledgements in trie
+		for ((port, channel, sequence), ack) in Acknowledgements::<T>::iter() {
+			let channel_id =
+				channel_id_from_bytes(channel).map_err(|_| Error::<T>::DecodingError)?;
+			let port_id = port_id_from_bytes(port).map_err(|_| Error::<T>::DecodingError)?;
+			let sequence = ibc::core::ics04_channel::packet::Sequence::from(sequence);
+
+			let ack_path = AcksPath { port_id, channel_id, sequence };
+
+			let mut ack_key = prefix.clone();
+			let ack_path_str = format!("{}", ack_path);
+			ack_key.extend_from_slice(ack_path_str.as_bytes());
+			inputs.push((ack_key.encode(), ack));
+		}
+
+		// Insert packet receipts in trie
+		for ((port, channel, sequence), receipt) in PacketReceipt::<T>::iter() {
+			let channel_id =
+				channel_id_from_bytes(channel).map_err(|_| Error::<T>::DecodingError)?;
+			let port_id = port_id_from_bytes(port).map_err(|_| Error::<T>::DecodingError)?;
+			let sequence = ibc::core::ics04_channel::packet::Sequence::from(sequence);
+
+			let receipt_path = ReceiptsPath { port_id, channel_id, sequence };
+
+			let mut receipt_key = prefix.clone();
+			let receipt_path_str = format!("{}", receipt_path);
+			receipt_key.extend_from_slice(receipt_path_str.as_bytes());
+			inputs.push((receipt_key.encode(), receipt))
+		}
+
+		Ok(sp_io::trie::blake2_256_root(inputs, sp_core::storage::StateVersion::V0))
+	}
+
+	pub(crate) fn extract_ibc_state_root() -> Result<Vec<u8>, Error<T>> {
+		let root = Self::build_ibc_state_root()?;
+		Ok(root.as_bytes().to_vec())
+	}
+
+	pub fn generate_raw_proof(keys: Vec<Vec<u8>>) -> Result<Vec<u8>, Error<T>> {
+		let keys = keys.iter().collect::<Vec<_>>();
+		let root = Self::build_ibc_state_root()?;
+		let db = sp_trie::MemoryDB::<BlakeTwo256>::default();
+		sp_trie::generate_trie_proof::<sp_trie::LayoutV0<BlakeTwo256>, _, _, _>(&db, root, keys)
+			.map(|proof| proof.encode())
+			.map_err(|_| Error::<T>::ProofGenerationError)
+	}
```

Comment:
This will unfortunately generate an empty proof, seeing as the db is empty. Anyways, Since this will be called in rpc, i think its fine to have a method that simply returns all the data that was committed at a specific height and do proof generation in native, instead of the runtime.